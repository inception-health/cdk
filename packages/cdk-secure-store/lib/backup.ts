import { aws_backup, aws_dynamodb, aws_kms } from "aws-cdk-lib";
import { Construct } from "constructs";

export interface SecureStoreBackupProps {
  backupPlanId?: string;
  encryptionKey: aws_kms.IKey;
  table: aws_dynamodb.ITableV2;
  tableName: string;
}

export interface SecureStoreBackupResources {
  backupPlan: aws_backup.IBackupPlan | undefined;
  backupVault: aws_backup.BackupVault | undefined;
}

export class SecureStoreBackup extends Construct {
  public readonly cdk: SecureStoreBackupResources;

  private readonly props: SecureStoreBackupProps;

  constructor(scope: Construct, id: string, props: SecureStoreBackupProps) {
    super(scope, id);

    // Initialize the properties
    this.props = props;
    this.cdk = {} as SecureStoreBackupResources;

    // Create the backup resources
    if (this.props.backupPlanId) {
      this.cdk.backupPlan = aws_backup.BackupPlan.fromBackupPlanId(
        this,
        "backupPlan",
        this.props.backupPlanId,
      );
    } else {
      // Create a backup vault
      this.cdk.backupVault = new aws_backup.BackupVault(this, "backupVault", {
        backupVaultName: `secure-store-${this.props.tableName}-backup-vault`,
        encryptionKey: this.props.encryptionKey,
        blockRecoveryPointDeletion: true,
        // TODO: Review if we want to make lock configuration by default.
        // lockConfiguration
      });

      // Create the backup plan
      this.cdk.backupPlan =
        aws_backup.BackupPlan.dailyWeeklyMonthly7YearRetention(
          this,
          `secure-store-${this.props.tableName}-backup-plan`,
          this.cdk.backupVault,
        );

      (this.cdk.backupPlan as aws_backup.BackupPlan).addSelection(
        `secure-store-${this.props.tableName}-backup-plan-selection`,
        {
          resources: [
            aws_backup.BackupResource.fromDynamoDbTable(this.props.table),
          ],
        },
      );
    }
  }
}
