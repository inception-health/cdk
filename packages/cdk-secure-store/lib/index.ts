export type { SecureStoreProps } from "./store";
export { SecureStore } from "./store";
