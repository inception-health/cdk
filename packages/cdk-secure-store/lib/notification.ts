import {
  aws_events,
  aws_events_targets,
  aws_kms,
  aws_logs,
  aws_sns,
} from "aws-cdk-lib";
import { Construct } from "constructs";

export interface SecureStoreNotificationProps {
  /**
   * The SNS Topic to send notifications to.
   */
  destination: string;
  /**
   * Enable CloudWatch Logs for the EventBridge rule.
   *
   * @defaultValue false
   */
  enableLogs?: boolean;
  /**
   * Enable or disable the notification.
   *
   * @defaultValue true
   */
  enabled?: boolean;
  /**
   * The encryption key to use for encrypting the sns topic and the logs
   *
   * @remarks
   * The encryption key must be created in the same account and region as the
   * DynamoDB table. It also must have the following policy attached to it:
   * ```
   * {
   *   "Sid": "Allow access to CloudWatch Log",
   *   "Effect": "Allow",
   *   "Principal": {
   *       "Service": "logs.REGION.amazonaws.com"
   *   },
   *   "Action": [
   *       "kms:Encrypt",
   *       "kms:Decrypt",
   *       "kms:ReEncrypt*",
   *       "kms:GenerateDataKey*",
   *       "kms:DescribeKey"
   *   ],
   *   "Resource": "*",
   *   "Condition": {
   *       "ArnLike": {
   *           "kms:EncryptionContext:aws:logs:arn": "arn:aws:logs:REGION:ACCOUNT_ID:*"
   *       }
   *   }
   * }
   *```
   */
  encryptionKey: aws_kms.IKey;
  /**
   * The secure store table name
   */
  sourceTable: string;
}

const defaultProps: Partial<SecureStoreNotificationProps> = {
  enabled: true,
  enableLogs: true,
};

interface SecureStoreNotificationResources {
  bus: aws_events.IEventBus;
  logs: aws_logs.LogGroup;
  rule: aws_events.Rule;
  topic: aws_sns.ITopic;
}

/**
 * `Notification` is a construct that creates a notification for specific
 *  DynamoDB events. The notification is sent to the provided SNS topic.
 *
 * At the moment, the notification is sent for the following events:
 * - CreateTable
 * - DeleteTable
 * - UpdateTable
 * - DeleteBackup
 * - RestoreTableFromBackup
 * - RestoreTableToPointInTime
 *
 * @remarks
 * For the full list of events that trigger a notification, see the [DynamoDB
 * documentation](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/logging-using-cloudtrail.html).
 */
export class SecureStoreNotification extends Construct {
  public readonly cdk: SecureStoreNotificationResources;

  // !Private Properties
  private readonly props: SecureStoreNotificationProps;

  // !Constructor
  constructor(
    scope: Construct,
    id: string,
    props: SecureStoreNotificationProps,
  ) {
    super(scope, id);

    // Initialize the properties
    this.props = { ...defaultProps, ...props };
    this.cdk = {} as SecureStoreNotificationResources;

    // Create Resources
    this.setupTopic();

    this.setupBus();

    this.setupRule();

    this.setupLogs();
  }

  // !Private Methods
  private setupTopic() {
    this.cdk.topic = aws_sns.Topic.fromTopicArn(
      this,
      "topic",
      this.props.destination,
    );
  }

  private setupBus() {
    this.cdk.bus = aws_events.EventBus.fromEventBusName(
      this,
      "eventBus",
      "default",
    );
  }

  private setupRule() {
    this.cdk.rule = new aws_events.Rule(this, "rule", {
      ruleName: `${this.props.sourceTable}-dynamodb--monitoring`,
      description: `Send system notifications related to the ${this.props.sourceTable} table.`,
      enabled: this.props.enabled,
      eventBus: this.cdk.bus,
      eventPattern: {
        source: ["aws.dynamodb"],
        detailType: ["AWS API Call via CloudTrail"],
        detail: {
          eventSource: ["dynamodb.amazonaws.com"],
          eventName: [
            "CreateTable",
            "DeleteTable",
            "UpdateTable",
            "DeleteBackup",
            "RestoreTableFromBackup",
            "RestoreTableToPointInTime",
          ],
          requestParameters: {
            tableName: [this.props.sourceTable],
          },
        },
      },
    });

    // Add the target to the rule
    this.cdk.rule.addTarget(new aws_events_targets.SnsTopic(this.cdk.topic));
  }

  private setupLogs() {
    if (!this.props.enableLogs) return;

    // New Cloudwatch Log Group to deliver EventBridge Events to.
    this.cdk.logs = new aws_logs.LogGroup(this, "logs", {
      logGroupName: `/inception/devops/${this.props.sourceTable}-dynamodb--monitoring`,
      retention: aws_logs.RetentionDays.TWO_WEEKS,
      encryptionKey: this.props.encryptionKey,
    });

    // Define rule that will deliver EventBridge events to a Cloudwatch Log Group
    this.cdk.rule.addTarget(
      new aws_events_targets.CloudWatchLogGroup(this.cdk.logs),
    );
  }
}
