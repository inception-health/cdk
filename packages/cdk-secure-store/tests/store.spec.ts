import {
  App,
  Stack,
  aws_kms,
  aws_dynamodb,
  CfnElement,
  aws_iam,
} from "aws-cdk-lib";
import { Match, Template } from "aws-cdk-lib/assertions";

import { SecureStore } from "../lib/store";

describe("Secure Store", () => {
  let stack: Stack;
  let encryptionKey: aws_kms.IKey;

  const devOpsTopicArn = "arn:aws:sns:us-east-1:123456789012:topic";
  const tableName = "MyTable";

  beforeEach(() => {
    const app = new App();
    stack = new Stack(app, "TestStack", {
      env: {
        region: "us-east-1",
        account: "123456789012",
      },
    });
    encryptionKey = aws_kms.Key.fromKeyArn(
      stack,
      "MyKey",
      "arn:aws:kms:us-east-1:123456789012:key/12345678-1234-1234-1234-123456789012",
    );
  });

  // ======================
  // General
  // ======================
  describe("General Behavior", () => {
    it("should match the default snapshot", () => {
      new SecureStore(stack, "MySecureStore", {
        tableName,
        encryptionKey: encryptionKey,
      });

      // then
      expect(Template.fromStack(stack)).toMatchSnapshot();
    });

    it("should create a SecureStoreDB table with the specified name", () => {
      new SecureStore(stack, "MySecureStore", {
        tableName,
        encryptionKey: encryptionKey,
      });

      // then
      const template = Template.fromStack(stack);

      template.hasResourceProperties("AWS::DynamoDB::GlobalTable", {
        TableName: tableName,
      });
    });

    it("should create a DynamoDB table with the specified encryption key", () => {
      const keyArn = "arn:aws:kms:us-east-1:123456789012:key/12345678";
      const encryptionKey = aws_kms.Key.fromKeyArn(stack, "MyKey-2", keyArn);

      new SecureStore(stack, "TestDynamo", {
        tableName,
        partitionKey: {
          name: "id",
          type: aws_dynamodb.AttributeType.STRING,
        },
        encryptionKey: encryptionKey,
      });

      const template = Template.fromStack(stack);

      template.hasResourceProperties("AWS::DynamoDB::GlobalTable", {
        SSESpecification: {
          SSEEnabled: true,
          SSEType: "KMS",
        },
        Replicas: [
          Match.objectLike({
            Region: "us-east-1",
            SSESpecification: {
              KMSMasterKeyId: keyArn,
            },
          }),
        ],
      });
    });

    it("it should have backups enabled", () => {
      // When
      const store = new SecureStore(stack, "MySecureStore", {
        tableName,
        encryptionKey: encryptionKey,
        // WHEN
        backupEnabled: true,
      });

      // Then
      expect(store.components.backup.cdk.backupPlan).toBeDefined();
      expect(store.components.backup.cdk.backupVault).toBeDefined();
      Template.fromStack(stack).resourceCountIs("AWS::Backup::BackupPlan", 1);
      Template.fromStack(stack).resourceCountIs("AWS::Backup::BackupVault", 1);
    });

    it("should have backup plan associated with the custom vault", () => {
      // When
      const store = new SecureStore(stack, "MySecureStore", {
        tableName,
        encryptionKey: encryptionKey,
        // WHEN
        backupEnabled: true,
      });

      // Then
      expect(store.components.backup.cdk.backupVault?.backupVaultArn).toBe(
        store.components.backup.cdk.backupVault?.backupVaultArn,
      );

      // Then
      const template = Template.fromStack(stack);

      const vaultName = stack.getLogicalId(
        store.components.backup.cdk.backupVault?.node
          .defaultChild as CfnElement,
      );

      template.hasResourceProperties("AWS::Backup::BackupPlan", {
        BackupPlan: {
          BackupPlanRule: Match.arrayEquals([
            {
              Lifecycle: {
                DeleteAfterDays: 35,
              },
              RuleName: "Daily",
              ScheduleExpression: "cron(0 5 * * ? *)",
              TargetBackupVault: {
                "Fn::GetAtt": [vaultName, "BackupVaultName"],
              },
            },
            {
              Lifecycle: {
                DeleteAfterDays: 90,
              },
              RuleName: "Weekly",
              ScheduleExpression: "cron(0 5 ? * SAT *)",
              TargetBackupVault: {
                "Fn::GetAtt": [vaultName, "BackupVaultName"],
              },
            },
            {
              Lifecycle: {
                DeleteAfterDays: 2555,
                MoveToColdStorageAfterDays: 90,
              },
              RuleName: "Monthly7Year",
              ScheduleExpression: "cron(0 5 1 * ? *)",
              TargetBackupVault: {
                "Fn::GetAtt": [vaultName, "BackupVaultName"],
              },
            },
          ]),
        },
      });
    });

    it("should have DevOps notifications when enabled", () => {
      // When
      const store = new SecureStore(stack, "MySecureStore", {
        tableName,
        encryptionKey: encryptionKey,
        devOpsTopicArn: devOpsTopicArn,
      });

      // Then
      expect(store.components.devops).toBeDefined();
      Template.fromStack(stack).resourceCountIs("AWS::Events::Rule", 1);
    });

    it("should have GSI when defined", () => {
      new SecureStore(stack, "MySecureStore", {
        tableName,
        encryptionKey: encryptionKey,
        // WHEN
        globalSecondaryIndexes: [
          {
            indexName: "MyGSI",
            partitionKey: {
              name: "id",
              type: aws_dynamodb.AttributeType.STRING,
            },
          },
        ],
      });

      // THEN
      Template.fromStack(stack).hasResourceProperties(
        "AWS::DynamoDB::GlobalTable",
        {
          GlobalSecondaryIndexes: [
            {
              IndexName: "MyGSI",
              KeySchema: [
                {
                  AttributeName: "id",
                  KeyType: "HASH",
                },
              ],
              Projection: {
                ProjectionType: "ALL",
              },
            },
          ],
        },
      );
    });

    it("should have LSI when defined", () => {
      new SecureStore(stack, "MySecureStore", {
        tableName,
        encryptionKey: encryptionKey,
        // WHEN
        localSecondaryIndexes: [
          {
            indexName: "MyGSI",
            sortKey: {
              name: "id",
              type: aws_dynamodb.AttributeType.STRING,
            },
          },
        ],
      });

      // THEN
      Template.fromStack(stack).hasResourceProperties(
        "AWS::DynamoDB::GlobalTable",
        {
          LocalSecondaryIndexes: [
            {
              IndexName: "MyGSI",
              KeySchema: [
                {
                  AttributeName: "PK",
                  KeyType: "HASH",
                },
                {
                  AttributeName: "id",
                  KeyType: "RANGE",
                },
              ],
              Projection: {
                ProjectionType: "ALL",
              },
            },
          ],
        },
      );
    });

    it("should raise exception when an encryption key is not specified", () => {
      // Given
      const devOpsTopicArn = "arn:aws:sns:us-east-1:123456789012:topic";
      const tableName = "MyTable";

      // When
      expect(() => {
        new SecureStore(stack, "MySecureStore", {
          tableName,
          devOpsTopicArn: devOpsTopicArn,
          // WHEN
          encryptionKey: undefined as any,
        });
      }).toThrowError("An encryption key is required for the table.");
    });

    it("should raise exception when a partition key is not specified", () => {
      // Given
      const devOpsTopicArn = "arn:aws:sns:us-east-1:123456789012:topic";
      const tableName = "MyTable";

      // When
      expect(() => {
        new SecureStore(stack, "MySecureStore", {
          tableName,
          encryptionKey: encryptionKey,
          devOpsTopicArn: devOpsTopicArn,
          // WHEN
          partitionKey: undefined as any,
        });
      }).toThrowError("The property partitionKey is required.");
    });
  });

  // ======================
  // Granting Permissions
  // ======================
  describe("Granting Permissions", () => {
    it("should grant read data permissions", () => {
      // Given
      const secureStore = new SecureStore(stack, "MySecureStore", {
        tableName,
        encryptionKey: encryptionKey,
      });
      const role = new aws_iam.Role(stack, "TestRole", {
        assumedBy: new aws_iam.ServicePrincipal("lambda.amazonaws.com"),
      });

      // WHEN
      secureStore.grantReadData(role);

      // Expect
      const logicalTableId = stack.getLogicalId(
        secureStore.cdk.table.node.defaultChild as CfnElement,
      );

      const logicalRoleId = stack.getLogicalId(
        role.node.defaultChild as CfnElement,
      );

      // THEN
      Template.fromStack(stack).hasResourceProperties("AWS::IAM::Policy", {
        PolicyDocument: {
          Statement: Match.arrayWith([
            {
              Action: ["kms:Decrypt", "kms:DescribeKey"],
              Effect: "Allow",
              Resource: encryptionKey.keyArn,
            },
            {
              Action: [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:DescribeTable",
              ],
              Effect: "Allow",
              Resource: [
                { "Fn::GetAtt": [logicalTableId, "Arn"] },

                {
                  "Fn::Join": [
                    "",
                    [{ "Fn::GetAtt": [logicalTableId, "Arn"] }, "/index/*"],
                  ],
                },
              ],
            },
          ]),
          Version: "2012-10-17",
        },
        Roles: [{ Ref: logicalRoleId }],
      });
    });

    it("should grant write data permissions", () => {
      // Given
      const secureStore = new SecureStore(stack, "MySecureStore", {
        tableName,
        encryptionKey: encryptionKey,
      });
      const role = new aws_iam.Role(stack, "TestRole", {
        assumedBy: new aws_iam.ServicePrincipal("lambda.amazonaws.com"),
      });

      // WHEN
      secureStore.grantWriteData(role);

      // Expect
      const logicalTableId = stack.getLogicalId(
        secureStore.cdk.table.node.defaultChild as CfnElement,
      );

      const logicalRoleId = stack.getLogicalId(
        role.node.defaultChild as CfnElement,
      );

      // THEN
      Template.fromStack(stack).hasResourceProperties("AWS::IAM::Policy", {
        PolicyDocument: {
          Statement: Match.arrayWith([
            {
              Action: [
                "kms:Decrypt",
                "kms:DescribeKey",
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
              ],
              Effect: "Allow",
              Resource: encryptionKey.keyArn,
            },
            {
              Action: [
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              Effect: "Allow",
              Resource: [
                { "Fn::GetAtt": [logicalTableId, "Arn"] },

                {
                  "Fn::Join": [
                    "",
                    [{ "Fn::GetAtt": [logicalTableId, "Arn"] }, "/index/*"],
                  ],
                },
              ],
            },
          ]),
          Version: "2012-10-17",
        },
        Roles: [{ Ref: logicalRoleId }],
      });
    });

    it("should grant read/write data permissions", () => {
      // Given
      const secureStore = new SecureStore(stack, "MySecureStore", {
        tableName,
        encryptionKey: encryptionKey,
      });
      const role = new aws_iam.Role(stack, "TestRole", {
        assumedBy: new aws_iam.ServicePrincipal("lambda.amazonaws.com"),
      });

      // WHEN
      secureStore.grantReadWriteData(role);

      // Expect
      const logicalTableId = stack.getLogicalId(
        secureStore.cdk.table.node.defaultChild as CfnElement,
      );

      const logicalRoleId = stack.getLogicalId(
        role.node.defaultChild as CfnElement,
      );

      // THEN
      Template.fromStack(stack).hasResourceProperties("AWS::IAM::Policy", {
        PolicyDocument: {
          Statement: Match.arrayWith([
            {
              Action: [
                "kms:Decrypt",
                "kms:DescribeKey",
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
              ],
              Effect: "Allow",
              Resource: encryptionKey.keyArn,
            },
            {
              Action: [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              Effect: "Allow",
              Resource: [
                { "Fn::GetAtt": [logicalTableId, "Arn"] },

                {
                  "Fn::Join": [
                    "",
                    [{ "Fn::GetAtt": [logicalTableId, "Arn"] }, "/index/*"],
                  ],
                },
              ],
            },
          ]),
          Version: "2012-10-17",
        },
        Roles: [{ Ref: logicalRoleId }],
      });
    });
  });
});
