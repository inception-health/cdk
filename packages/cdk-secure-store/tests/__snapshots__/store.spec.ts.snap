// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Secure Store General Behavior should match the default snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "MySecureStorealarmsalarmerrorsCC0F67CA": {
      "Properties": {
        "AlarmName": {
          "Fn::Join": [
            "",
            [
              "DynamoDB ",
              {
                "Ref": "MySecureStoretable0562C06B",
              },
              " System Errors",
            ],
          ],
        },
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "DatapointsToAlarm": 1,
        "EvaluationPeriods": 6,
        "Metrics": [
          {
            "Expression": "getitem + putitem + deleteitem + updateitem + transactwriteitems + transactgetitems + batchgetitem + batchwriteitem + query + scan",
            "Id": "expr_1",
            "Label": "Sum of errors across all operations",
          },
          {
            "Id": "getitem",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "Operation",
                    "Value": "GetItem",
                  },
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "MySecureStoretable0562C06B",
                    },
                  },
                ],
                "MetricName": "SystemErrors",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "putitem",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "Operation",
                    "Value": "PutItem",
                  },
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "MySecureStoretable0562C06B",
                    },
                  },
                ],
                "MetricName": "SystemErrors",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "deleteitem",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "Operation",
                    "Value": "DeleteItem",
                  },
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "MySecureStoretable0562C06B",
                    },
                  },
                ],
                "MetricName": "SystemErrors",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "updateitem",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "Operation",
                    "Value": "UpdateItem",
                  },
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "MySecureStoretable0562C06B",
                    },
                  },
                ],
                "MetricName": "SystemErrors",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "transactwriteitems",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "Operation",
                    "Value": "TransactWriteItems",
                  },
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "MySecureStoretable0562C06B",
                    },
                  },
                ],
                "MetricName": "SystemErrors",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "transactgetitems",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "Operation",
                    "Value": "TransactGetItems",
                  },
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "MySecureStoretable0562C06B",
                    },
                  },
                ],
                "MetricName": "SystemErrors",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "batchgetitem",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "Operation",
                    "Value": "BatchGetItem",
                  },
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "MySecureStoretable0562C06B",
                    },
                  },
                ],
                "MetricName": "SystemErrors",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "batchwriteitem",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "Operation",
                    "Value": "BatchWriteItem",
                  },
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "MySecureStoretable0562C06B",
                    },
                  },
                ],
                "MetricName": "SystemErrors",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "query",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "Operation",
                    "Value": "Query",
                  },
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "MySecureStoretable0562C06B",
                    },
                  },
                ],
                "MetricName": "SystemErrors",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "scan",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "Operation",
                    "Value": "Scan",
                  },
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "MySecureStoretable0562C06B",
                    },
                  },
                ],
                "MetricName": "SystemErrors",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 0,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "MySecureStorealarmsalarmthrottle67AB3E62": {
      "Properties": {
        "AlarmName": {
          "Fn::Join": [
            "",
            [
              "DynamoDB ",
              {
                "Ref": "MySecureStoretable0562C06B",
              },
              " Throttled",
            ],
          ],
        },
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "DatapointsToAlarm": 1,
        "EvaluationPeriods": 6,
        "Metrics": [
          {
            "Expression": "m1 + m2",
            "Id": "expr_1",
            "Label": {
              "Fn::Join": [
                "",
                [
                  "DynamoDB ",
                  {
                    "Ref": "MySecureStoretable0562C06B",
                  },
                  " Reads/Writes Throttled",
                ],
              ],
            },
          },
          {
            "Id": "m1",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "MySecureStoretable0562C06B",
                    },
                  },
                ],
                "MetricName": "ReadThrottleEvents",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "m2",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "MySecureStoretable0562C06B",
                    },
                  },
                ],
                "MetricName": "WriteThrottleEvents",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 0,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "MySecureStorebackupbackupVault114398DB": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AccessPolicy": {
          "Statement": [
            {
              "Action": [
                "backup:DeleteRecoveryPoint",
                "backup:UpdateRecoveryPointLifecycle",
              ],
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "BackupVaultName": "secure-store-MyTable-backup-vault",
        "EncryptionKeyArn": "arn:aws:kms:us-east-1:123456789012:key/12345678-1234-1234-1234-123456789012",
      },
      "Type": "AWS::Backup::BackupVault",
      "UpdateReplacePolicy": "Retain",
    },
    "MySecureStorebackupsecurestoreMyTablebackupplan2A315DAA": {
      "Properties": {
        "BackupPlan": {
          "BackupPlanName": "secure-store-MyTable-backup-plan",
          "BackupPlanRule": [
            {
              "Lifecycle": {
                "DeleteAfterDays": 35,
              },
              "RuleName": "Daily",
              "ScheduleExpression": "cron(0 5 * * ? *)",
              "TargetBackupVault": {
                "Fn::GetAtt": [
                  "MySecureStorebackupbackupVault114398DB",
                  "BackupVaultName",
                ],
              },
            },
            {
              "Lifecycle": {
                "DeleteAfterDays": 90,
              },
              "RuleName": "Weekly",
              "ScheduleExpression": "cron(0 5 ? * SAT *)",
              "TargetBackupVault": {
                "Fn::GetAtt": [
                  "MySecureStorebackupbackupVault114398DB",
                  "BackupVaultName",
                ],
              },
            },
            {
              "Lifecycle": {
                "DeleteAfterDays": 2555,
                "MoveToColdStorageAfterDays": 90,
              },
              "RuleName": "Monthly7Year",
              "ScheduleExpression": "cron(0 5 1 * ? *)",
              "TargetBackupVault": {
                "Fn::GetAtt": [
                  "MySecureStorebackupbackupVault114398DB",
                  "BackupVaultName",
                ],
              },
            },
          ],
        },
      },
      "Type": "AWS::Backup::BackupPlan",
    },
    "MySecureStorebackupsecurestoreMyTablebackupplansecurestoreMyTablebackupplanselection180E0B92": {
      "Properties": {
        "BackupPlanId": {
          "Fn::GetAtt": [
            "MySecureStorebackupsecurestoreMyTablebackupplan2A315DAA",
            "BackupPlanId",
          ],
        },
        "BackupSelection": {
          "IamRoleArn": {
            "Fn::GetAtt": [
              "MySecureStorebackupsecurestoreMyTablebackupplansecurestoreMyTablebackupplanselectionRole0B5FF01B",
              "Arn",
            ],
          },
          "Resources": [
            {
              "Fn::GetAtt": [
                "MySecureStoretable0562C06B",
                "Arn",
              ],
            },
          ],
          "SelectionName": "secure-store-MyTable-backup-plan-selection",
        },
      },
      "Type": "AWS::Backup::BackupSelection",
    },
    "MySecureStorebackupsecurestoreMyTablebackupplansecurestoreMyTablebackupplanselectionRole0B5FF01B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "backup.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSBackupServiceRolePolicyForBackup",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MySecureStoretable0562C06B": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "PK",
            "AttributeType": "S",
          },
          {
            "AttributeName": "SK",
            "AttributeType": "S",
          },
          {
            "AttributeName": "GSI1PK",
            "AttributeType": "S",
          },
          {
            "AttributeName": "GSI1SK",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "GlobalSecondaryIndexes": [
          {
            "IndexName": "GSI1",
            "KeySchema": [
              {
                "AttributeName": "GSI1PK",
                "KeyType": "HASH",
              },
              {
                "AttributeName": "GSI1SK",
                "KeyType": "RANGE",
              },
            ],
            "Projection": {
              "ProjectionType": "ALL",
            },
          },
        ],
        "KeySchema": [
          {
            "AttributeName": "PK",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "SK",
            "KeyType": "RANGE",
          },
        ],
        "Replicas": [
          {
            "ContributorInsightsSpecification": {
              "Enabled": true,
            },
            "DeletionProtectionEnabled": true,
            "GlobalSecondaryIndexes": [
              {
                "ContributorInsightsSpecification": {
                  "Enabled": true,
                },
                "IndexName": "GSI1",
              },
            ],
            "PointInTimeRecoverySpecification": {
              "PointInTimeRecoveryEnabled": true,
            },
            "Region": "us-east-1",
            "SSESpecification": {
              "KMSMasterKeyId": "arn:aws:kms:us-east-1:123456789012:key/12345678-1234-1234-1234-123456789012",
            },
          },
        ],
        "SSESpecification": {
          "SSEEnabled": true,
          "SSEType": "KMS",
        },
        "TableName": "MyTable",
      },
      "Type": "AWS::DynamoDB::GlobalTable",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
